lib/logstash/filters/memcached.rb

```diff
@@ -67,28 +68,50 @@ class LogStash::Filters::Memcached < LogStash::Filters::Base
   # NOTE: in Memcached, a value of 0 (default) means "never expire"
   config :ttl, :validate => :number, :default => 0
 
+  # Tags the event on failure. This can be used in later analysis.
+  config :tag_on_failure, :validate => :string, :default => "_memcached_failure"
+
   public
 
   attr_reader :cache
 
   def register
-    if @ttl < 0
-      logger.error("ttl cannot be negative")
-      fail("invalid ttl: cannot be negative")
+    raise(LogStash::ConfigurationError, "'ttl' option cannot be negative") if @ttl < 0
+
+    @memcached_hosts = validate_connection_hosts
+    @memcached_options = validate_connection_options
+    begin
+      @cache = new_connection(@memcached_hosts, @memcached_options)
+    rescue => e
+      logger.error("failed to connect to memcached", hosts: @memcached_hosts, options: @memcached_options, message: e.message)
+      fail("failed to connect to memcached")
     end
-
-    @cache = establish_connection
-  end # def register
+    @connected = true
+  end
 
   def filter(event)
-    set_success = do_set(event)
-    get_success = do_get(event)
+    unless @connected
+      return unless reconnect(@memcached_hosts, @memcached_options)
+    end
 
-    filter_matched(event) if (set_success || get_success)
-  end # def filter
+    begin
+      set_success = do_set(event)
+      get_success = do_get(event)
+      filter_matched(event) if (set_success || get_success)
+    rescue Dalli::NetworkError, Dalli::RingError => e
+      event.tag(@tag_on_failure)
+      logger.error("memcached communication error",  hosts: @memcached_hosts, options: @memcached_options, message: e.message)
+      close
+    end
+  end
 
   def close
+    @connected = false
```

Comment:
we _may_ want to do this in mutual exclusion too, to ensure we don't close a cache that had just been opened by another thread as a part of connection recovery.

Reply:
as per my main thread comment, I totally overlooked thread safety. I will correct that.